Userland/Libraries/LibGfx/ImageFormats/JBIG2Loader.cpp

```diff
@@ -16,6 +16,207 @@
 
 namespace Gfx {
 
+namespace JBIG2 {
+
+// Table E.1 – Qe values and probability estimation process
+// See also E.1.2 Coding conventions and approximations
+// and E.2.5 Probability estimation.
+struct QeEntry {
+    u16 qe;          // Sub-interval for the less probable symbol.
+    u16 nmps;        // Next index if the more probable symbol is decoded
+    u16 nlps;        // Next index if the less probable symbol is decoded
+    u16 switch_flag; // See second-to-last paragraph in E.1.2.
+};
+constexpr auto qe_table = to_array<QeEntry>({
+    { 0x5601, 1, 1, 1 },
+    { 0x3401, 2, 6, 0 },
+    { 0x1801, 3, 9, 0 },
+    { 0x0AC1, 4, 12, 0 },
+    { 0x0521, 5, 29, 0 },
+    { 0x0221, 38, 33, 0 },
+    { 0x5601, 7, 6, 1 },
+    { 0x5401, 8, 14, 0 },
+    { 0x4801, 9, 14, 0 },
+    { 0x3801, 10, 14, 0 },
+    { 0x3001, 11, 17, 0 },
+    { 0x2401, 12, 18, 0 },
+    { 0x1C01, 13, 20, 0 },
+    { 0x1601, 29, 21, 0 },
+    { 0x5601, 15, 14, 1 },
+    { 0x5401, 16, 14, 0 },
+    { 0x5101, 17, 15, 0 },
+    { 0x4801, 18, 16, 0 },
+    { 0x3801, 19, 17, 0 },
+    { 0x3401, 20, 18, 0 },
+    { 0x3001, 21, 19, 0 },
+    { 0x2801, 22, 19, 0 },
+    { 0x2401, 23, 20, 0 },
+    { 0x2201, 24, 21, 0 },
+    { 0x1C01, 25, 22, 0 },
+    { 0x1801, 26, 23, 0 },
+    { 0x1601, 27, 24, 0 },
+    { 0x1401, 28, 25, 0 },
+    { 0x1201, 29, 26, 0 },
+    { 0x1101, 30, 27, 0 },
+    { 0x0AC1, 31, 28, 0 },
+    { 0x09C1, 32, 29, 0 },
+    { 0x08A1, 33, 30, 0 },
+    { 0x0521, 34, 31, 0 },
+    { 0x0441, 35, 32, 0 },
+    { 0x02A1, 36, 33, 0 },
+    { 0x0221, 37, 34, 0 },
+    { 0x0141, 38, 35, 0 },
+    { 0x0111, 39, 36, 0 },
+    { 0x0085, 40, 37, 0 },
+    { 0x0049, 41, 38, 0 },
+    { 0x0025, 42, 39, 0 },
+    { 0x0015, 43, 40, 0 },
+    { 0x0009, 44, 41, 0 },
+    { 0x0005, 45, 42, 0 },
+    { 0x0001, 45, 43, 0 },
+    { 0x5601, 46, 46, 0 },
+});
+
+ErrorOr<ArithmeticDecoder> ArithmeticDecoder::initialize(ReadonlyBytes data, Context context)
+{
+    ArithmeticDecoder decoder { data };
+    decoder.CX = context;
+    decoder.INITDEC();
+    return decoder;
+}
+
+bool ArithmeticDecoder::get_next_bit()
+{
+    // Useful for comparing to Table H.1 – Encoder and decoder trace data.
+    // dbg("I={} MPS={} A={:#x} C={:#x} CT={} B={:#x}", I(CX), MPS(CX), A, C, CT, B());
```

Comment:
Commented out dbglns? Pls remove or add a verbose debug flag for dbgln_if

Reply:
It's for every single bit that's decided. Having a dbgln_if() is not useful. It _is_ useful for the test case this adds and was super useful while debugging this. I can remove this if you want, but it is truly useful. ("But it works now?!" Yes, but it's the naive impl and maybe I want to rewrite it to be faster at some point.)

Reply:
(For context, see  https://www.itu.int/rec/T-REC-T.88-201808-I page 164. It has a table with the expected internal states for that test datastream, and this prints that table, more or less.)

Reply:
hm, well if it's more of an implementation note than actual debugging, I won't die on that hill. :shrug: